src/stdatamodels/fits_support.py

```diff
@@ -644,10 +713,52 @@ def from_fits_asdf(hdulist,
     Wrap asdf call to extract optional arguments
     """
     ignore_missing_extensions = kwargs.pop('ignore_missing_extensions')
-    return fits_embed.AsdfInFits.open(hdulist,
-                                      ignore_version_mismatch=ignore_version_mismatch,
-                                      ignore_unrecognized_tag=ignore_unrecognized_tag,
-                                      ignore_missing_extensions=ignore_missing_extensions)
+
+    try:
+        asdf_extension = hdulist[_ASDF_EXTENSION_NAME]
+    except (KeyError, IndexError, AttributeError):
+        # This means there is no ASDF extension
+        return asdf.AsdfFile(
+            ignore_version_mismatch=ignore_version_mismatch,
+            ignore_unrecognized_tag=ignore_unrecognized_tag,
+        )
+
+    generic_file = generic_io.get_file(io.BytesIO(asdf_extension.data), mode="rw")
+    return asdf.open(
+        generic_file,
+        ignore_version_mismatch=ignore_version_mismatch,
+        ignore_unrecognized_tag=ignore_unrecognized_tag,
+        ignore_missing_extensions=ignore_missing_extensions,
+        tree_mapper=partial(_convert_fits_arrays, hdulist)
```

Comment:
Is this the only change required to the ASDF API to enable this PR? Is there a way to reimplement ASDF-in-FITS without this? One idea would be to implement the conversion in a Converter (which has precedence over old style types). This would require adding block/block_manager support to Converters and would likely benefit from a 'accept/reject' system like mentioned here: https://github.com/asdf-format/asdf/issues/1263